Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -57,119 +59,164 @@ void Mixer::DoState(PointerWrap& p)
-  // remember fractional offset
+    const float x = pi * (i - frac);
+    float weight = 0.53836f + 0.46164f * std::cos(x / sinc_window_width);
+    if (std::abs(x) >= 1e-6f)
+      weight *= std::sin(x * sinc_ratio) / x;
+    else
+      weight *= sinc_ratio;
 
-  float aid_sample_rate =
-      FIXED_SAMPLE_RATE_DIVIDEND / static_cast<float>(m_input_sample_rate_divisor);
-  if (consider_framelimit && emulationspeed > 0.0f)
-  {
-    float numLeft = static_cast<float>(((indexW - indexR) & INDEX_MASK) / 2);
+    r_sample += weight * r;
+    l_sample += weight * l;
+  }
 
-    u32 low_watermark = (FIXED_SAMPLE_RATE_DIVIDEND * timing_variance) /
-                        (static_cast<u64>(m_input_sample_rate_divisor) * 1000);
-    low_watermark = std::min(low_watermark, MAX_SAMPLES / 2);
+  return std::make_pair(l_sample, r_sample);
+}
 
-    m_numLeftI = (numLeft + m_numLeftI * (CONTROL_AVG - 1)) / CONTROL_AVG;
-    float offset = (m_numLeftI - low_watermark) * CONTROL_FACTOR;
-    if (offset > MAX_FREQ_SHIFT)
-      offset = MAX_FREQ_SHIFT;
-    if (offset < -MAX_FREQ_SHIFT)
-      offset = -MAX_FREQ_SHIFT;
+// Executed from sound stream thread
+void Mixer::MixerFifo::Mix(s16* samples, u32 num_samples, double target_speed)
+{
+  // Cache access in non-volatile variable. No race conditions are here, as indexW only increases
+  // (which is fine), and this is the only place indexR is written to.
+  const u32 indexW = m_indexW.load();  // Write index in circular buffer
+  u32 indexR = m_indexR.load();        // Read index in circular buffer
+
+  const float l_volume = float(m_LVolume.load()) / 256.0f;
+  const float r_volume = float(m_RVolume.load()) / 256.0f;
+
+  auto clamp = [](s32 input) -> s16 { return static_cast<s16>(std::clamp(input, -32768, 32767)); };
+
+  const double ratio = (target_speed * FIXED_SAMPLE_RATE_DIVIDEND) /
+                       (m_input_sample_rate_divisor * m_mixer->m_sampleRate);
+
+  // These are timing markers. We want to keep the audio buffer around the low watermark.
+  const s32 direct_latency = m_mixer->m_config_direct_latency;
+  const s32 low_watermark = std::min<s32>(MAX_SAMPLES, direct_latency * FIXED_SAMPLE_RATE_DIVIDEND /
+                                                           (m_input_sample_rate_divisor * 1000));
+
+  // Calculate the number of samples remaining in the buffer
+  const s32 sinc_window_width = m_mixer->m_config_sinc_window_width;
+  s32 remaining_samples = s32((indexW - indexR) & INDEX_MASK) - sinc_window_width - 2;
+
+  // Calculate the speed at which to play the requested samples in order to maintain the correct
+  // synchronization with the video stream. Done by slightly adjusting the pitch of the audio.
+  const double requested_samples = m_multiplier * num_samples * ratio;
+  double multiplier = (remaining_samples - low_watermark) / requested_samples;
+  multiplier = 1.0 + CONTROL_EFFORT * (multiplier - 1.0);
+  multiplier = std::clamp(multiplier, MIN_PITCH_SHIFT, MAX_PITCH_SHIFT);
```

Comment:
Do I understand correctly this merely slows down the audio stream by less than 2%? What scenarios exactly does this help with?